### ISSUE TITLE
Crash when "unexpected end of file" encountered.

### ISSUE THREAD

**Issue by RandssINC** [JohnnyFFM/miner-burst#2]
Would get this error occasionally but miner would not crash with previous version.
Appears to be the AVX version that does this.
Also, Would it be possible to change "discarded D/L" message to any other color besides red?

**Comment by joyrider3774** [comment]
i'm having similar issue with miner-911_AVX2.exe it shows a red message "error p1 readfile error =" but no error code, sometimes it does show it, and the windows error watchdog is triggerd and it catches the unhandled exception.

I have some problems with usb 3.0 hub and a failing disk as well (i think) and sometimes i get errors on disk according to blagominer but it never crashed with triggering the windows watchdog

windows event logs reports this
```
Naam van toepassing met fout: miner-911_AVX2.exe, versie: 0.0.0.0, tijdstempel: 0x5b12d04a
Naam van module met fout: miner-911_AVX2.exe, versie: 0.0.0.0, tijdstempel: 0x5b12d04a
Uitzonderingscode: 0xc0000409
Foutmarge: 0x00000000000537c1
Id van proces met fout: 0x40e4
Starttijd van toepassing met fout: 0x01d3fc37eb16bde4
Pad naar toepassing met fout: C:\tools\miner-burst-1.170603\miner-911_AVX2.exe
Pad naar module met fout: C:\tools\miner-burst-1.170603\miner-911_AVX2.exe
Rapport-id: 9f75b086-fc0f-430b-924f-53286a970438
Volledige pakketnaam met fout: 
Relatieve toepassings-id van pakket met fout:
 ```

0xc0000409 means STATUS_STACK_BUFFER_OVERRUN.

i'm getting this on a windows 10 build 1709 (so does not have april update yet) 64 bit proffessional using i5-5250u intel cpu

i switched back to miner-v1.170901_AVX2

**Comment by JohnnyFFM** [comment]
Hi, many thanks for feedback. The 911 is still work in progress. I've pushed an update this morning and would be interested if this is fixing your issues. With the color of the discarded D/L message, I totally agree. Red means something goes wrong and that's not the case for a discarded D/L. Next time I'm in the code I'll amended it to dark green.

**Comment by joyrider3774** [comment]
Is the release binary updated to reflect these latest changes ? if so i'll try it when i get home. I usually get quite a lot of those error messages so should be able to easily test it to see if it's fixed

**Comment by JohnnyFFM** [comment]
Yes, I've updated the binaries! Thanks for testing!

**Comment by mpwsh** [comment]
I'm receiving the same error with version 911. After the error 4 of my drives disappear and i have to restart the computer, lol.

**Comment by JohnnyFFM** [comment]
Did you test the 911 beta2?

**Comment by mpwsh** [comment]
Yes, i did the test with beta2 version. (blagominer-v1.170911_beta2_POC2)

**Comment by JohnnyFFM** [comment]
Ok, this one is giving me a hard time, running it on 4 machines sas, usb sata all mixed, no issues at all. I‘ll keep on investigating. Also last thing you see is the p1 readfileerror?

**Comment by joyrider3774** [comment]
i can confirm it stall crashes / triggers the windows watchdog with beta2 version, i just tried it as well, exception code is the same

edit: i do not always see the p1 readfileerror i had it crash without displaying the message as well
the problem probably only arrises if there are errors detected during scanning of the file. This is just a wild guess but what happens if you deliberatly create an unfinished plot file and use it in blago miner ? not sure it will trigger the error but could be it will (like kill the plotting program half way through the plot file so you know the plot file is possible not finished / broken)

**Comment by joyrider3774** [comment]
i've got a crash dump file for debugging, am uploading it to my google drive and will share in moment by editing this message

edit: here's the dump file generated using wer using beta 2 available in releases following these steps: https://blogs.msdn.microsoft.com/chaun/2013/11/12/steps-to-catch-a-simple-crash-dump-of-a-crashing-process/

this link should work if not let me know
https://drive.google.com/file/d/1gUrOcmZG_TkPaE67wdfu_QfK1bQKkecH/view?usp=sharing

i know it's possible to use this for debugging but i can not tell you how exactly as i have no experiance with it

edit2: check https://msdn.microsoft.com/en-us/library/windows/desktop/ee416349(v=vs.85).aspx it seems you should be able to open them in visual studio

edit3: not sure if it's needed but perhaps a crashdump based on a binary containing debug information will yield more info / function names in the stack trace ? (i tried opening it in the free version of visual studio and it said there were no debugging symbols present)

**Comment by mpwsh** [comment]
Sorry, i think its important to mention that the drives having this problem are filled with PoC1 plot files.
I'm testing the miner with the only poc2 drive i have available now, and it's not throwing any errors.

Just for clarification, if i start the miner with PoC1 and PoC2 drives mixed i get the error.
And if i start the miner with PoC2 drives no errors are shown (so far)

**Comment by JohnnyFFM** [comment]
Thanks for all the hints. For the drive with POC1 files only: can you narrow it down to a single drive and provide me with the filelist? I would then quikly recreate your file structure with dummy files. Dont need your id though.

**Comment by joyrider3774** [comment]
i downloaded the source code and i managed to build the release version, use a remote debugger to debug from my pc running the miner / having the problem but unfortunatly same error about no debugging symbols and when i try to create a debug version i get a lot of unresolved linker errors. Since i'm not used to visual studio c/c++ i'm not sure how i can get it to compile a debug version (i'm a delphi programmer). I think if i could have a debug version build i can probably remote debug and view a stacktrace or debug correctly 

here are some of the linker errors but not all
```1>miner.obj : error LNK2001: unresolved external symbol __imp_exit
1>miner.obj : error LNK2001: unresolved external symbol __imp_memmove
1>miner.obj : error LNK2001: unresolved external symbol __imp__invalid_parameter
1>miner.obj : error LNK2001: unresolved external symbol __imp_strpbrk
1>miner.obj : error LNK2001: unresolved external symbol __imp___acrt_iob_func
1>miner.obj : error LNK2001: unresolved external symbol __imp_fflush
1>miner.obj : error LNK2001: unresolved external symbol __imp_fclose
1>msvcrtd.lib(utility.obj) : error LNK2001: unresolved external symbol _cexit
1>msvcrtd.lib(utility.obj) : error LNK2001: unresolved external symbol __C_specific_handler
1>msvcrtd.lib(delete_scalar.obj) : error LNK2001: unresolved external symbol _free_dbg
1>msvcrtd.lib(utility_desktop.obj) : error LNK2001: unresolved external symbol terminate
1>msvcrtd.lib(throw_bad_alloc.obj) : error LNK2001: unresolved external symbol __std_exception_copy
1>msvcrtd.lib(throw_bad_alloc.obj) : error LNK2001: unresolved external symbol __std_exception_destroy
1>msvcrtd.lib(throw_bad_alloc.obj) : error LNK2001: unresolved external symbol _CxxThrowException
```

**Comment by joyrider3774** [comment]
managed to get a call stack but not sure it's helpfull

```
 	miner.exe!abort() Line 77	C++
 	miner.exe!terminate() Line 58	C++
 	[Inline Frame] miner.exe!std::thread::{dtor}() Line 57	C++
>	miner.exe!work_i(const unsigned __int64 local_num) Line 1771	C++
 	[External Code]	
 	[Inline Frame] miner.exe!invoke_thread_procedure(unsigned int(*)(void *)) Line 91	C++
 	miner.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter) Line 115	C++
 	[External Code]	

```

it comes from here https://github.com/JohnnyFFM/miner-burst/blob/performance/miner/miner.cpp#L1771

then it goes to thread line 57 which says

```
~thread() _NOEXCEPT
		{	// clean up
		if (joinable())
			_XSTD terminate(); <-- unhandled exception visual studio shows here
		}

```

**Comment by mpwsh** [comment]
just FYI, i reinstalled my SATA controller's driver and the issue seems to be gone (so far). I've been mining for 2 hours with 911 beta2 and received no errors.

**Comment by JohnnyFFM** [comment]
@joyrider3774 Same issue on by side with the DEBUG, did not manage to run it so far . Will look into this today, seems like an issue with pdcurses. Will also create some borked plot files on my side and see if I can somehow reproduce the error.

**Comment by joyrider3774** [comment]
@JohnnyFFM i managed to get that last stack trace by not changing the build from release to debug but by editing the release config, linker options and just selecting to Generate debug info (i had it set to optimize for debugging) it still compiled and it seems to show a stacktrace then, as well as highlight the line where execuption occured in visual studio itselve. I think that was the only change i did to get that stacktrace while it crashed

**Comment by JohnnyFFM** [comment]
Good news, I've managed to have both debug and release building working.  I've just started from scratch, creating an empty C++ project in VC2017, added pdcurses and injected Blago's code step by step. So got rid of a lot of garbage as well. What I'm going to do (probably tonight, latest tomorrow morning) is to create another repository with the refactored code and add some information on how to build it. I'll spinoff the 911 development to this new repository and abandon this one here at 901. 

@joyrider3774 : Would be cool if you could then try to build the spinoff repo and debug this nasty bug.

**Comment by JohnnyFFM** [comment]
And for the meantime, attached the compiled version of the spinoff just for someone to confirm that the bug is still around =).  update, found a bug and fixed it, beta4 now.  
[blagominer911b4.zip](https://github.com/JohnnyFFM/miner-burst/files/2076693/blagominer911b4.zip)

**Comment by joyrider3774** [comment]
@JohnnyFFM i just tested 911 b4 (zip file above) and crash still happens. Anytime i get some error while reading plot file the crash occurs (this time i did see the ERROR P1 Readfile message) from one of my failing disks. Tonight i'll try to create a borked plot file and see if it triggers the error each time or at least more frequently (it depends where data is being read from the plot file /diks) in my case. Some blocks no error at all occur but some blocks i do get the errors and in case of these beta versions a crash.

I'll try out the debugging but i'm in noway familiar with visual studio / C++ but i'll give it a shot.

**Comment by JohnnyFFM** [comment]
Oki. Will debug again and simulate a borked file in code and maybe will try to mine a usb disk and unplug it while reading...

**Comment by joyrider3774** [comment]
@JohnnyFFM created borked plot file does not seem to work with me i tried everything like modifying contents with hex editor, make read access denied, killing xplotter.exe during plotting etc.

However disconnecting an usb drive with a plot file while the miner is reading from it does make it happen with me on my main pc. I used an usb disk with a 500 gig plot file so i had time to disconnect the usb drive while the miner was reading from it

same exception code is written then in windows event log (0xc0000409) if i use the old miner-v1.170901_AVX2 and do the same the miner does not crash like this and just shows an error message in the console window.

My guess is you need some file I/O error for it to trigger

**Comment by joyrider3774** [comment]
@JohnnyFFM  commenting out this line https://github.com/JohnnyFFM/miner-burst/blob/performance/miner/miner.cpp#L1771 (basically the break statement) does not make the miner crash but it shows a lot of unexpected end of file red errors then (at least when disconnecting the drive) so that break statement is the cause somehow. 

not sure if that's a valid solution because i really get a lot of unexpected errors but now you should be able to simulate it with you as well by disconnecting usb drives while reading from it and have something to start with 

edit: i tried again without the break statement, removed drive -> no crash but a lot of unexpected end of file errors -> reattached drives while miner was still running (because it did not crash now) ->happily mined further on next block without having to restart the miner or getting an app crash / unhandled exception

**Comment by joyrider3774** [comment]
@JohnnyFFM i found the culprit you create the read thread but you need to detach it before using break in case err is true because it was never joined to main thread it causes app crashes. the read thread is still joinable and goes out of scope due to the break. In the destructor of the thread there is a verifications that checks If the thread is joinable, and if it is std::terminate is called

this also fixes it with me without 100's of unexpected end of file errors, it only showed 1 now as it did in older versions
```
//Join threads
hash.join();
if (err)
{
  if (read.joinable())
    read.detach();
  break;
}
read.join();
if (cont) continue;
count += 1;				


```

i found this by searching google the 1st item here seems relevant in this case:
http://www.acodersjourney.com/2017/08/top-20-cplusplus-multithreading-mistakes/

**Comment by JohnnyFFM** [comment]
Awesome! Many thanks for tracking down this bug! Excellent work. I'll create the new executables tomorrow and commit everything. 

Will also need to check if the unexpected end of file won't cause a crash, but that's a quick thing, I'll just take a dummy file where the size is not a multiple of 2^18 and rename it like a plot file. 

Many thanks again! And I'm still a little bit ashamed having committet the number one multithreadding mistake =)

**Comment by joyrider3774** [comment]
@JohnnyFFM no problem.

Btw even though the app crash is fixed, it seems error logging is somehow a bit diffrent compared to older version (901). I took 2 screenshots top one is using old miner, bottom one is with my fix applied on same block

https://imgur.com/a/uOV5qHy

so might need to look into this as well, unless it's normal

**Comment by joyrider3774** [comment]
sorry @JohnnyFFM  but i had let the miner.exe with my fix run all night and it seems there might be another bug somewhere i had another app crash for another reason. This time an access violation was the cause. Luckily i still had the generation of crash dumps active and the miner.Exe contained some debug symbols.

here's the place where it happened and the call stack in a screenshot. So this causes app crashes as well somehow but i have not looked into it. It might also have something todo with the weird error logging i mentioned in previous post / screenshot and more specifically with the iter variable when an error happens

https://imgur.com/a/PusXHEb

the call stack was 
```
 	miner.exe!std::operator+<char,std::char_traits<char>,std::allocator<char> >(const char * _Right, const std::basic_string<char,std::char_traits<char>,std::allocator<char> > &) Line 2329	C++
>	miner.exe!th_read(void * ifile, const unsigned __int64 start, const unsigned __int64 MirrorStart, bool * const cont, unsigned __int64 * const bytes, const t_files * const iter, bool * const flip, bool p2, const unsigned __int64 i, const unsigned __int64 stagger, unsigned __int64 * const cache_size_local, char * const cache, char * const MirrorCache) Line 1494	C++
 	[External Code]	
 	[Inline Frame] miner.exe!invoke_thread_procedure(unsigned int(*)(void *)) Line 91	C++
 	miner.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter) Line 115	C++
 	[External Code]	

```

i switched that line of code back to like it is for poc2 read and also was in original sources:

```
wprintw(win_main, "error P1 ReadFile. code = %llu\n", GetLastError(), 0);
```

now it correctly shows the code (getlast error) again. I'm letting it run for the rest of the night and will check tommorrow to see if it has crashed again or not

edit more crashes (access violations) happen and not always on the same place but always related to th_read code, well all my access violations were inside th_read when i check stack traces but not always on the same line, not sure what's happening

**Comment by JohnnyFFM** [comment]
Hi, I've just committed all changes. I fixed one more bug passing the file iterator to th read and th hash, might be the cause for the crashes. 
[miner911-AVX-b5.zip](https://github.com/JohnnyFFM/miner-burst/files/2078930/miner911-AVX-b5.zip)

**Comment by JohnnyFFM** [comment]
I've just uploaded a "clean" version of Blagominer 911 into a new repository. The debug function works for me with this one. I've uploaded the two pdcurses libraries as well, I keep them in c:\pdcurses\lib. BTW, I'm compiling the miner with v140 (10% faster than v141).

**Comment by joyrider3774** [comment]
@JohnnyFFM I'll get those sources when i get home then.

I've installed the beta 5, i had it crash once already, it was almost immediatly after starting it, i started it again when i was on train and 20 minutes later it is still running. It does not crash at "unexpected end of file" errors anymore with me, but there might still be some problem in it. I did notice i did not have a lot of File I/O errors now so maybe the "new" crashes are still related to when file i/o errors occur somehow, but not immediatly.

I'll let it run further and check back from now and then, i'm at work now so can't do much atm, but perhaps tonight i'll look further but i'm not sure if i might be able to fix it, last night the stack pointed on 3 diffrent places inside th_read when the access violation happened. Could still be thread related but not sure

edit: 2nd attempt and it's been running for over an hour atm without a crash so things have definatly improved but it did crash once, as i see the dump file from beta 5

edit2: crashed after about 1.5 hours.

**Comment by JohnnyFFM** [comment]
I reviewed the code again and found three more things:
(1) the "if (cont) continue;" for the first read and the final "if (err) break;" don't make  sense as it is already outside the relevant loop. 
(2) If cont happens, the next hashing should be skipped.
(3) if (cont) continue;	count += 1; should first increase count, then continue next. 

Additionally I have to re-think what happens if an err happens (error when hashing, typically unexpected end of file). Old code skips the next read, but when dual threading I don't have this info yet and will start the next read. This next read should fail as well. 

beta6 
[blagominer911b6-AVX2.zip](https://github.com/JohnnyFFM/miner-burst/files/2079802/blagominer911b6-AVX2.zip)

Went one step further and amended the dual threading. I precheck if the hashing would fail and dont start the read if so. Additionally fixed one bug with the cache and unoptimized plot files. 

beta7
[blagominer911b7-AVX.zip](https://github.com/JohnnyFFM/miner-burst/files/2080216/blagominer911b7-AVX.zip)

**Comment by joyrider3774** [comment]
@JohnnyFFM i'm testing beta 7 now, hasn't crashed yet but it's only been running for 15 minutes. What i can say is that your changes fixed the error logging output, it is back to like it was in original version like only showing the error once and not twice. The 2nd error log entry probably happend from still reading after gotten error in previous versions. Here's a screenshot https://imgur.com/a/Mh4eF3G

If it crashes i'll edit the post here and if it does not like say after a day i'll mention it as well

edit: it's been running for about 7 hours without a crash atm

**Comment by JohnnyFFM** [comment]
Looks like we we have a release candidate 😄 Guess if it survives the night it is good to go! 
So this will be a release weekend, today morning xplotter 1.3, tomorrow the 911 blagominer and  PocConverter 2.0.

**Comment by joyrider3774** [comment]
It's been running for about 17 hours without any crash atm, my nuc (where the miner is runnig on) is never turned of so i'll leave beta 7 on it, it's indeed looking good now 

edit: 28 hours & no crash

**Comment by JohnnyFFM** [comment]
@joyrider3774 That's good news! Just created a release. Thanks again for all the help!
Best, 
Johnny
